test/test_mypy.py

```diff
@@ -127,65 +138,212 @@ def test_default_document_type(self) -> None:
+            ]
+        )
+        self.assertTrue(len(list(result)))
+
+
+class TestDecode(unittest.TestCase):
+    def test_bson_decode(self) -> None:
+        doc = {"_id": 1}
+        bsonbytes = encode(doc)
+        rt_document: Dict[str, Any] = decode(bsonbytes)
+        assert rt_document["_id"] == 1
+        rt_document["foo"] = "bar"
+
+        class MyDict(Dict[str, Any]):
+            def foo(self):
+                return "bar"
+
+        codec_options = CodecOptions(document_class=MyDict)
+        bsonbytes2 = encode(doc, codec_options=codec_options)
+        rt_document2 = decode(bsonbytes2, codec_options=codec_options)
+        assert rt_document2.foo() == "bar"
+
+        codec_options2 = CodecOptions(document_class=RawBSONDocument)
+        bsonbytes3 = encode(doc, codec_options=codec_options2)
+        rt_document3 = decode(bsonbytes2, codec_options=codec_options2)
+        assert rt_document3.raw
+
+    def test_bson_decode_all(self) -> None:
+        doc = {"_id": 1}
+        bsonbytes = encode(doc)
+        bsonbytes += encode(doc)
+        rt_documents: List[Dict[str, Any]] = decode_all(bsonbytes)
+        assert rt_documents[0]["_id"] == 1
+        rt_documents[0]["foo"] = "bar"
+
+        class MyDict(Dict[str, Any]):
+            def foo(self):
+                return "bar"
+
+        codec_options2 = CodecOptions(MyDict)
+        bsonbytes2 = encode(doc, codec_options=codec_options2)
+        bsonbytes2 += encode(doc, codec_options=codec_options2)
+        rt_documents2 = decode_all(bsonbytes2, codec_options2)
+        assert rt_documents2[0].foo() == "bar"
+
+        codec_options3 = CodecOptions(RawBSONDocument)
+        bsonbytes3 = encode(doc, codec_options=codec_options3)
+        bsonbytes3 += encode(doc, codec_options=codec_options3)
+        rt_documents3 = decode_all(bsonbytes3, codec_options3)
+        assert rt_documents3[0].raw
+
+    def test_bson_decode_iter(self) -> None:
+        doc = {"_id": 1}
+        bsonbytes = encode(doc)
+        bsonbytes += encode(doc)
+        rt_documents: Iterator[Dict[str, Any]] = decode_iter(bsonbytes)
+        assert next(rt_documents)["_id"] == 1
+        next(rt_documents)["foo"] = "bar"
+
+        class MyDict(Dict[str, Any]):
+            def foo(self):
+                return "bar"
+
+        codec_options2 = CodecOptions(MyDict)
+        bsonbytes2 = encode(doc, codec_options=codec_options2)
+        bsonbytes2 += encode(doc, codec_options=codec_options2)
+        rt_documents2 = decode_iter(bsonbytes2, codec_options2)
+        assert next(rt_documents2).foo() == "bar"
+
+        codec_options3 = CodecOptions(RawBSONDocument)
+        bsonbytes3 = encode(doc, codec_options=codec_options3)
+        bsonbytes3 += encode(doc, codec_options=codec_options3)
+        rt_documents3 = decode_iter(bsonbytes3, codec_options3)
+        assert next(rt_documents3).raw
+
+    def make_tempfile(self, content: bytes) -> Any:
+        fileobj = tempfile.TemporaryFile()
```

Comment:
I suggest adding `self.addCleanup(fileobj.close)` to close the file.

Reply:
Done